### ISSUE TITLE
About multi-gpu  and the details of training

### ISSUE THREAD

**Issue by morning4346** [PRIS-CV/Mutual-Channel-Loss#3]
I can't run this code when I want to use multi-GPU.
I only change os.environ["CUDA_VISIBLE_DEVICES"] = "2,3"
Is there any question?DO I need change another code?

**Comment by dongliangchang** [comment]
Yes, if you want to ues multi-GPU, you need to use the DataParallel function.
To make it easy, I have change the code of CUB-200-2011.py to make it can handle multi-GPU, you can see the changes in line 220.

**Comment by morning4346** [comment]
It works,thank you very much.

**Comment by morning4346** [comment]
you train the net about 300 epoch,It really spend long times if I only use two GPUs（one epoch 5 minutes）,Does it really need? if I change the lr whether it can reach the accuracy , and decrease the epoch.like 100 epoch.

**Comment by dongliangchang** [comment]
Just try, and if you can share your results to here, it will be very useful to others. 
many thanks.

**Comment by morning4346** [comment]
ok,I also have another question.
as your code describe,you use 224*224 image size,I got 68.0%acc,
but when I want to use 448*448 image size,I changed the data process,and 600*14*14,and lr,but it does not work,the train acc convergence  Rapidly,but the test acc can only reach 20%,I don't know why ,whether it has some place I don't change,or some other reasons.

**Comment by morning4346** [comment]
I think it must change the cnum as your paper said,some class have two channels,while others have three channels,but I don't know have to change,can you share your code to me ?as you train the vgg16 with the 448*448 image size and resnet50 with 448*448 size train process,thanks very much.

**Comment by morning4346** [comment]
of course,I think it must change the my_pooling layer,can you share your code?

**Comment by dongliangchang** [comment]
> 
> 
> I think it must change the cnum as your paper said,some class have two channels,while others have three channels,but I don't know have to change,can you share your code to me ?as you train the vgg16 with the 448_448 image size and resnet50 with 448_448 size train process,thanks very much.

"some class have two channels,while others have three channels"  is the situation  that trained model with the **pre-trained** VGG16 or ResNet50. When we **TRAINED FROM SCRATCH**, we can change the  total  feature channels = ξ * the categories nums.

**Comment by dongliangchang** [comment]
> 
> 
> ok,I also have another question.
> as your code describe,you use 224_224 image size,I got 64.1%acc,of course,it can't get you acc(65.8%). but when I want to use 448_448 image size,I changed the data process,and 600_14_14,and lr,but it does not work,the train acc convergence Rapidly,but the test acc can only reach 20%,I don't know why ,whether it has some place I don't change,or some other reasons.

I will try, and release the code. TRAINED FROM SCRATCH with 448 input size.

**Comment by dongliangchang** [comment]
>  my_pooling layer

No need~

**Comment by dongliangchang** [comment]
> 
> 
> I think it must change the cnum as your paper said,some class have two channels,while others have three channels,but I don't know have to change,can you share your code to me ?as you train the vgg16 with the 448_448 image size and resnet50 with 448_448 size train process,thanks very much.

If you use the pre-trained VGG16 or ResNet50, you also can use the 1*1 conv to change the channels.

**Comment by morning4346** [comment]
I want to know you result about pretrained vgg16 get 78.7%,and pretrained resnet50 get 87.3%,I think it has image size 448*448,and use pretrained vgg16 and resnet50,it does not need trained from scratch .
so If I want to reach your acc,can you share your code ?

**Comment by dongliangchang** [comment]
How to train with pre-trained model is somewhat tedious, requiring many modifications and not compatible with the current code. That's why the code for trained with pre-trained model  is not released.

**Comment by morning4346** [comment]
It would be very nice if you release  the code.

**Comment by dongliangchang** [comment]
> 
> 
> It would be very nice if you release the code.

The code has been released.

**Comment by Kurumi233** [comment]
Hello, here is a reimplementation version. The code of mcloss is integrated into a class, that can be easily embedded in trainning. Would you like to check it ? 
https://github.com/Kurumi233/Mutual-Channel-Loss
Thanks :)

**Comment by dongliangchang** [comment]
> 
> 
> Hello, here is a reimplementation version. The code of mcloss is integrated into a class, that can be easily embedded in trainning. Would you like to check it ?
> https://github.com/Kurumi233/Mutual-Channel-Loss
> Thanks :)

This repo integrate the MC-Loss into a class, easy to use in practice. many thanks.
